src/tests/suite/daos_nvme_recovery.c

```diff
@@ -119,9 +119,165 @@ nvme_recov_1(void **state)
+	 * Verify log_mask in server.yaml file, It should be 'DEBUG' to verify
+	 * different state of NVMe drives. Skip the test if log_mask is not
+	 * set to DEBUG.
+	 */
+	D_ALLOC(server_config_file, 512);
+	D_ALLOC(log_file, 1024);
+	for (i = 0; i < ndisks; i++) {
+		if (devices[i].rank == 1) {
+			rc = get_server_config(devices[i].host,
+					       server_config_file);
+			assert_int_equal(rc, 0);
+			print_message("server_config_file = %s\n",
+				      server_config_file);
+
+			get_server_log_file(devices[i].host,
+					    server_config_file, log_file);
+			rc = verify_server_log_mask(devices[i].host,
+						    server_config_file,
+						    "DEBUG");
+			if (rc) {
+				print_message("Log Mask != DEBUG in %s.\n",
+					      server_config_file);
+				skip();
+			}
+		}
+	}
+	print_message("LOG FILE = %s\n", log_file);
+
+	/** Prepare records **/
+	if (arg->pool.pool_info.pi_nnodes < 2)
+		obj_class = DAOS_OC_R1S_SPEC_RANK;
+	else
+		obj_class = DAOS_OC_R2S_SPEC_RANK;
+
+	oid = dts_oid_gen(obj_class, 0, arg->myrank);
+	oid = dts_oid_set_rank(oid, 1);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+	memset(data_buf, 'a', 100);
+
+	/** Insert record **/
+	print_message("Insert single record with 100 extents\n");
+	insert_single_with_rxnr("dkey", "akey", 0, data_buf,
+				1, 100, DAOS_TX_NONE, &req);
+
+	/**
+	*Set single device for rank1 to faulty.
+	*/
+	for (i = 0; i < ndisks; i++) {
+		if (devices[i].rank == 1) {
+			print_message(
+				"NVMe with UUID=%s on host=%s\" set to Faulty\n",
+				DP_UUID(devices[i].device_id),
+				devices[i].host);
+			rc = dmg_storage_set_nvme_fault(dmg_config_file,
+							devices[i].host,
+				devices[i].device_id, 1);
+			assert_int_equal(rc, 0);
+			break;
+		}
+	}
+	sleep(60);
+
+	/**
+	* Verify Rank1 device state change from NORMAL to FAULTY.
+	* Verify "FAULTY -> TEARDOWN" and "TEARDOWN -> OUT" device states found
+	* in server log.
+	*/
+	rc = dmg_storage_device_list(dmg_config_file, NULL, devices);
+	assert_int_equal(rc, 0);
+	for (i = 0; i < ndisks; i++) {
+		if (devices[i].rank == 1) {
+			assert_string_equal(devices[i].state, "\"FAULTY\"");
+
+			rc = verify_state_in_log(devices[i].host,
+						 log_file, "TEARDOWN -> OUT");
+			if (rc != 0) {
+				print_message(
+					"TEARDOWN -> OUT not found in log %s\n",
+					log_file);
+				assert_int_equal(rc, 0);
+			}
+
+			rc = verify_state_in_log(devices[i].host,
+						 log_file,
+						 "FAULTY -> TEARDOWN");
+			if (rc != 0) {
+				print_message(
+					"FAULTY -> TEARDOWN not found in %s\n",
+					log_file);
+				assert_int_equal(rc, 0);
+			}
+			break;
+		}
+
+	}
+
+	/** Lookup all the records and verify the content **/
+	print_message("Lookup and Verify all the records:\n");
+	lookup_single_with_rxnr("dkey", "akey", 0, fetch_buf,
+				1, 100, DAOS_TX_NONE, &req);
+	for (i = 0; i < 100; i++)
+		assert_memory_equal(&fetch_buf[i], "a", 1);
+
+	/*
+	 * FIXME: Add FAULTY disks back to the system, when feature available.
```

Comment:
this can be done once $dmg storage replace command is added

Reply:
Right